### PR TITLE
Fix optional colocation outweighing infinite stickiness

### PR DESCRIPTION
**Pull request by kgaillot**
This fixes the previously added colocation-vs-stickiness scheduler regression test.

Previously, chained optional colocations could result in an unnecessary ban for the primary resource when a negative score was involved. The first dependent's preferences would be incorporated at an attenuated factor, but the chained dependents' preferences would be incorporated at full-strength. Now, those are attentuated as well.

### CODE REVIEW

**Comment by kgaillot** [comment]
@nrwahl2 , I finally have a fix for what I was trying to fix, as opposed to issues I found while investigating! No rush

File: lib/pacemaker/pcmk_sched_colocation.c [hunk 1]
```diff
@@ -1474,9 +1473,11 @@ pcmk__add_colocated_node_scores(pe_resource_t *rsc, const char *log_id,
         flags |= pcmk__coloc_select_active;
 
         for (GList *iter = colocations; iter != NULL; iter = iter->next) {
-            pe_resource_t *other = NULL;
             pcmk__colocation_t *constraint = (pcmk__colocation_t *) iter->data;
 
+            pe_resource_t *other = NULL;
+            float other_factor = factor * constraint->score / (float) INFINITY;
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_sched_colocation.c:1479]: I can't think of any unwanted side effects for "this with" colocations (this commit is aimed at "with this"), and obviously regression passed; hopefully there aren't any

File: cts/scheduler/scores/colocation-vs-stickiness.scores [hunk 2]
```diff
@@ -57,7 +57,7 @@ pcmk__primitive_assign: FencingPass allocation score on rhel7-5: INFINITY
 pcmk__primitive_assign: dummy1 allocation score on rhel7-1: -INFINITY
 pcmk__primitive_assign: dummy1 allocation score on rhel7-2: -INFINITY
 pcmk__primitive_assign: dummy1 allocation score on rhel7-3: -INFINITY
-pcmk__primitive_assign: dummy1 allocation score on rhel7-4: -500
+pcmk__primitive_assign: dummy1 allocation score on rhel7-4: 0
```
Comment (nrwahl2) [inline cts/scheduler/scores/colocation-vs-stickiness.scores:60]: How did we lose the anti-colocation score? It should lose the battle compared to the stickiness in this test but I wonder if there are cases where it would make a difference.

**Review by nrwahl2 — APPROVED** [review]
Hey, look how simple this was after only six months of preparation!

This looks good to me. You said the choice seems somewhat arbitrary. This is just about the only way that I think makes sense. Before I got to the bottom paragraph of the main commit message, this is how I was thinking it ought to be done.

By multiplying the factors recursively, this gets us as we can reasonably get to having the effect of "tentatively assign B based on its 'with this' colocations, then assign A considering B's 'final' scores." (If B has "this with" colocations with some other resource D, then those probably don't get considered by the algorithm... but that's a different headache that we can deal with later, or never.)

The original approach of course didn't account for the fact that A only gives 1% of a crap about C, because A only cares about C via B and A only gives 1% of a crap about B.

The only other approach that feels halfway reasonable is to apply the "B with A" factor directly to all the chained colocations instead of multiplying the factor recursively. I think this is better though, for the reason listed two paragraphs ago.

For sanity, it might be good to test with non-infinite stickiness and make sure the result is still the same. It most likely is. It wouldn't necessarily need a regression test.


> The dependent's node preferences are added to the primary's, attenuated by a
factor equal to the colocation score divided by INFINITY. Thus, for a mandatory
colocation, the dependent's preferences are incorporated at full strength,

Hopefully there's other code that prevents a mandatory colocation from keeping the primary from running anywhere, if it's unable to run on any of the nodes where the dependent can run; or a mandatory anti-colocation from keeping the primary from running anywhere, if the primary and the dependent are both able to run only on the same node. In those cases I'd assume we'd somehow roll back the dependent preferences and try again.

File: cts/scheduler/scores/colocation-vs-stickiness.scores (hunk 2 above)
Comment (kgaillot) [reply cts/scheduler/scores/colocation-vs-stickiness.scores:60]: At the start, both group1 and group2 are active on rhel7-5. Not sure why that score ever got applied to 7-4 :)

Comment (nrwahl2) [reply cts/scheduler/scores/colocation-vs-stickiness.scores:60]: I ran `crm_simulate` on this input file. First `pcmk__group_assign()` runs on `group1` and its members and then `group2` and its members (probably correct although I would've expected `group2` first since it's the primary resource... must not matter for group_assign).

Then `pcmk__primitive_assign()` runs on each of `group2`'s members, and then each of `group1`'s members.

All the `pcmk__group_assign()` outputs had `INFINITY` for the current node `rhel7-5` and zeroes elsewhere. The `dummy2` resources got assigned first and chose `rhel7-4` due to anti-colocation with `dummy1`. Specifically, `dummy1` has an `INFINITY` preference for `rhel7-5` due to stickiness, so `-1 * INFINITY = -INFINITY`.

As a result, `dummy2` chose the other available node, `rhel7-4`. 

When `dummy1` resources got assigned, they anti-colocated with a score of -500.

After the fix, `dummy2` appropriately weights `dummy1`'s preference, so it gets a temporary score of `-0.0005 * INFINITY = -500` for `rhel7-5`. When its own preference of `INFINITY` is added, the result is `INFINITY`, and it chooses `rhel7-5` instead of incorrectly choosing `rhel7-4`.

Basically same thing then happens when assigning `dummy1`. It starts with an `INFINITY` score for `rhel7-5` due to stickiness. 500 gets subtracted from the score three times (for `dummy2a`, `dummy2b`, and `dummy2c`), but `INFINITY - 500 = INFINITY`.
